blackjax/tempered_smc.py

```diff
@@ -0,0 +1,251 @@
+    """ Current state for the tempered SMC algorithm. This encodes the current set of sampled particles
+    as well as the corresponding tempering parameter :math:`\lambda`
+    """
+    particles: PyTree
+    lamda: float
+
+
+class SMCInfo(NamedTuple):
+    """Additional information on the tempered SMC step.
+
+    weights: jnp.ndarray
+        The weights after the MCMC pass.
+    total_steps: int
+        Number of steps in the MCMC pass.
+    """
+    weights: jnp.ndarray
+    total_steps: int
+
+
+def kernel(potential_fn: Callable, mcmc_kernel_factory: Callable, new_mcmc_state: Callable, resampling_method: Callable,
+           mcmc_stop_criterion: Callable):
+    """Build a Tempered SMC step.
+
+    Parameters
+    ----------
+    potential_fn: Callable
+        A function that returns the potential energy of a chain at a given position.
+    mcmc_kernel_factory: Callable
+        A callable function that creates a kernel from `potential_function`
+    new_mcmc_state: Callable
+        How to create a new MCMC state for the MCMC routine to start
+    resampling_method: Callable
+        A random function that resamples generated particles based of weights
+    mcmc_stop_criterion: Callable
+        A function that takes all the MCMC states, the mcmc iteration count and returns False when the chain should stop
+
+    Returns
+    -------
+    A callable that takes a rng_key and a Pytree that contains the current state
+    of the chain and that returns a new state of the chain along with
+    information about the transition.
+    """
+
+    def one_step(rng_key: jnp.ndarray, state: SMCState, delta: float) -> Tuple[SMCState, SMCInfo]:
+        """
+
+        Parameters
+        ----------
+        rng_key: DeviceArray[int],
+            JAX PRNGKey for randomness
+        state: SMCState
+            Current state of the tempered SMC algorithm
+        delta: float
+            Increment for `state.lamda
+
+        Returns
+        -------
+        state: SMCState,
+            The new state of the tempered SMC algorithm
+         info: SMCInfo,
+            Additional information on the SMC step
+        """
+        n_particles = jax.tree_flatten(state.particles)[0][0].shape[0]
+
+        tempered_potential_fn = lambda pytree: -state.lamda * potential_fn(pytree)
+
+        step_mcmc_kernel = jax.vmap(mcmc_kernel_factory(tempered_potential_fn), in_axes=(0, 0))
+        rng_key, resampling_key = jax.random.split(rng_key, 2)
+
+        stop_criterion = lambda carry: mcmc_stop_criterion(carry[0], carry[1])
+
+        def mcmc_loop_body(carry):
+            i, mcmc_state, key = carry
+            key, body_key = jax.random.split(key, 2)
+            op_keys = jax.random.split(body_key, n_particles)
+            mcmc_state, _ = step_mcmc_kernel(op_keys, mcmc_state)
+            return i + 1, mcmc_state, key
+
+        initial_mcmc_states = jax.vmap(new_mcmc_state, in_axes=[0, None])(state.particles, tempered_potential_fn)
```

Comment:
We would get rid of a lot of code by letting `particles` be the MCMC state. Since all MCMC states have a `position` argument, we could safely pass `position.state` to every function that only needs the position.

So this line would go away.

Reply:
ok